open-banking-accelerator/components/com.wso2.openbanking.accelerator.gateway/src/main/java/com/wso2/openbanking/accelerator/gateway/executor/dcr/DCRExecutor.java

```diff
@@ -610,6 +250,409 @@ public void postProcessRequest(OBAPIRequestContext obapiRequestContext) {
+                        .concat(clientId), basicAuthHeader);
+                handleInternalServerError(obapiResponseContext, OpenBankingErrorCodes.REGISTRATION_UPDATE_ERROR);
+                return;
+            }
+            String applicationId = applicationSearchResponse.getAsJsonObject().get("list").getAsJsonArray().get(0)
+                    .getAsJsonObject().get(applicationIdParam).getAsString();
+
+            //get list of subscribed APIs
+            JsonElement subscribedAPIsResponse = callGet(urlMap.get(GatewayConstants.API_GET_SUBSCRIBED).toString(),
+                    GatewayConstants.BEARER_TAG.concat(token), "applicationId", applicationId);
+            if (subscribedAPIsResponse == null) {
+                log.error("Error while retrieving subscribed APIs");
+                //delete SP created to call dev portal REST APIs
+                callDelete(urlMap.get(GatewayConstants.IAM_DCR_URL).toString().concat("/")
+                        .concat(clientId), basicAuthHeader);
+                handleInternalServerError(obapiResponseContext, OpenBankingErrorCodes.REGISTRATION_UPDATE_ERROR);
+                return;
+            }
+            List<String> subscribedAPIIdList = new ArrayList<>();
+            for (JsonElement subscribedAPI : subscribedAPIsResponse.getAsJsonObject().get("list")
+                    .getAsJsonArray()) {
+                String apiId = subscribedAPI.getAsJsonObject().get("apiId").getAsString();
+                subscribedAPIIdList.add(apiId);
+            }
+
+            //get software statement from dcr app details
+            String softwareStatement = createdDCRAppDetails.has(OpenBankingConstants.SOFTWARE_STATEMENT) ?
+                    createdDCRAppDetails.get(OpenBankingConstants.SOFTWARE_STATEMENT).getAsString() : null;
+            if (StringUtils.isNotEmpty(softwareStatement)) {
+                final JsonArray subscribedAPIs = subscribedAPIsResponse.getAsJsonObject()
+                        .get("list").getAsJsonArray();
+                //check whether the ssa still contains the roles related to the subscribed APIs and unsubscribe if not
+                Optional.of(getRolesFromSSA(softwareStatement))
+                        .map(ssaRoles -> getUnAuthorizedAPIs(subscribedAPIs, regulatoryAPIs, ssaRoles))
+                        .flatMap(unAuthorizedApis -> unAuthorizedApis.stream()
+                                .map(unAuthorizedApi -> String.format("%s/%s",
+                                        urlMap.get(GatewayConstants.API_GET_SUBSCRIBED).toString(), unAuthorizedApi))
+                                .filter(endpoint -> {
+                                    try {
+                                        return !callDelete(endpoint, GatewayConstants.BEARER_TAG.concat(token));
+                                    } catch (OpenBankingException | IOException e) {
+                                        return false;
+                                    }
```

Comment:
Shall we move this block to a different function which returns a boolean?

Reply:
Resolved in [8657c32](https://github.com/wso2/financial-open-banking/pull/62/commits/8657c32a1adada4bcf5b652287b825b09a36ca96)